### ISSUE TITLE
Improve Binder support for Java records

### ISSUE THREAD

**Issue by mcollovati** [vaadin/flow#20318]
### Describe your motivation

Currently, `Binder` (and `BeanValidationBinder`) work with Java records, but with a couple of limitations:

* "virtual properties" are not handled (e.g. a method in the record class that computes a value at runtime, like `boolean adult() { return age >= 18; }`)
* all record components must be bound to a field, otherwise the `writeRecord()` operation fails

Java records could potentially be improved to support the mentioned cases by

* inspecting the record class for public methods with a not void return type and 0 arguments (side note: `Introspector` can be used to find `getXXXX` like properties on records too, although it would be a bit hacky to add getters like properties in a record)
* If not all record components are bound, there's probably still have a chance to find an additional constructor taking only part of the components.

### Describe the solution you'd like

Binder supports "virtual properties" and custom constructors for Java records

### Describe alternatives you've considered

Nope

### Additional context

`BeanBinderTest` can be replicated using records instead of a Java beans. It would already provide tests for the described uses cases.

**Comment by Legioth** [comment]
> If not all record components are bound, there's probably still have a chance to find an additional constructor taking only part of the components.

I don't think that would be enough. One specific example is if I have an `id` property in my record to be able to store it in the database. I would want `writeRecord` to return a record that contains the same `id` value that was in the value that I previously passed to `readRecord`.

**Comment by Legioth** [comment]
It might be weird if we keep unbound values around for `readRecord` / `readBean` in case you write back using `writeRecord` but not when writing back with `writeBean`. That got me thinking that one potential API design approach could be if `readRecord` returns some kind of "record binding" object that captures the originally read record. This new object would have its own `writeRecord` method that allows pulling defaults from the captured original record when creating the new record.

**Comment by mcollovati** [comment]
Could it be having a `R writeRecord(R)`  method a good compromise, somehow consistent with `writeBean(B)`? 
The original record with the ID is provided, and the method will merge changed values with the untouched from the given record.

**Comment by Legioth** [comment]
That might also be a good option. Do you @knoobie have any additional insight based on your use cases?

**Comment by knoobie** [comment]
I like the idea of Marco! It might be a bit tricky to "merge" the correct values but it could theoretically work. You might even call it changed bindings like the similar method for normal beans / dtos. 

Another alternative I'm thinking about is similar to Leif's idea that setRecord (is that even a thing?) and readRecord both store the "initial" record in the binder allowing to use it as base / compare new attributes with it.. technically (correct me if I'm wrong.. haven't tried it with records)  you could map the record attributes based on their field's name if the source code was compiled with -parameter which should be like the gold standard nowadays (especially with spring kinda also enforcing it to get good DX)

**Comment by mcollovati** [comment]
> you could map the record attributes based on their field's name if the source code was compiled with -parameter

The flag is probably not necessary. From the `Class` instance, you can get the `RecordComponent`s that also store the record attribute `name`.

**Comment by mcollovati** [comment]
The only thing I don't like that much is the need to store an instance/map of values for `readRecord`. This way, the Binder becomes "stateful" behaving differently from `readBean`

**Comment by knoobie** [comment]
While I agree with your stand point from a framework perspective.. I wonder what the users would expect and feels more "naturally" / easy to use. 

If I call "readRecord(R)" and have bound 2/3 of fields. I don't wanna lose the third value (and I don't wanna bloat my code with it) 

Later when I call "writeRecord()" - I would expect a new record based on the old values + newly changed one. Like 2/3 from bindings (or 1/2 if only one is changed) and the third from the "stored" record so that I can use it.

Similarly: when we talk about "writeRecord(R)" where I can supply the value from outside, I would expect the same "logic" like previously - with the only difference - that the external R would be the initial value / value provider for the missing / unchanged attributes.

**Comment by knoobie** [comment]
> Do you knoobie have any additional insight based on your use cases?

Coming back to Leif's original question: I might not be the perfect target audience:  I would also need for forField support because all my fields have to be bound programmically and not based on reflection / "magic" because of advanced validation rulesets and so on.  (Which of course would also need the record to be stored in the Binder)
